spmd/tensor/parallel/api.py

```diff
@@ -1,122 +1,195 @@
 # Copyright (c) Meta Platforms, Inc. and affiliates
 import torch
 import torch.nn as nn
-from typing import Sequence, Tuple
+from typing import Union, Dict
 from spmd.tensor import (
     distribute_module,
     distribute_tensor,
-    DTensor,
     Shard,
     Replicate,
     DeviceMesh,
-    Placement,
 )
 from spmd.tensor.parallel import TensorParallelMultiheadAttention
-from spmd.tensor.parallel.style import ParallelStyle, PairwiseParallel
+from spmd.tensor.parallel.style import PairwiseParallel, ParallelStyle
 from spmd.tensor.parallel.utils import _create_1d_device_mesh
 
 
-def replicate_input(
-    inputs: Sequence[torch.Tensor], device_mesh: DeviceMesh
-) -> Tuple[DTensor, ...]:
-    replicate = [Replicate()] * device_mesh.ndim
-    return tuple(
-        DTensor.from_local(tensor, device_mesh, replicate) for tensor in inputs
-    )
+__all__ = [
+    "parallelize_module",
+]
 
 
-def replicate_output(output: DTensor, device_mesh: DeviceMesh) -> torch.Tensor:
-    if isinstance(output, DTensor):
-        replicate = [Replicate()] * output.device_mesh.ndim
-        # TODO: can the output be left incontiguous?
-        return (
-            output.redistribute(output.device_mesh, replicate)
-            .to_local()
-            .contiguous()
-        )
+def parallelize_module(
+    module: nn.Module,
+    device_mesh: DeviceMesh,
+    parallelize_plan: Union[ParallelStyle, Dict[str, ParallelStyle]],
+    tp_mesh_dim: int = 0,
+) -> nn.Module:
+    """
+    The major API for tensor parallelism in PyTorch. We parallelize module
+    or sub_modules based on a parallelize_plan which contains the parallel_style
+    which indicates how user want the module or sub_module to be parallelized.
+    User can also specify different parallel_style per FQN. The API supports 2D
+    parallelism natively by accepting an n-dimension device_mesh and users just
+    need to specify the dimension where we perform tensor parallelism on.
+
+    Args:
+        module (nn.Module):
+            :class:``nn.Module`` object to be parallelized.
+        device_mesh (DeviceMesh):
+            :class:``DeviceMesh`` object which describes the mesh topology
+            of devices for the DTensor.
+        parallelize_plan (Union[ParallelStyle, Dict[str, ParallelStyle]]):
+            The plan used to parallelize the module. It can be either a
+            :class:``ParallelStyle`` object which contains how
+            we prepare input/output for Tensor Parallelism or it can be a
+            dict of FQN and its corresponding :class:``ParallelStyle`` object.
```

Comment:
we should say module FQN to avoid confusion with parameter FQN

Reply:
sure.